File: LeetCode/P232_ImplementQueueUsingStacks.java
/*
 * LeetCode 232 - Implement Queue using Stacks
 * https://leetcode.com/problems/implement-queue-using-stacks/
 *
 * Approach: Two stacks - "inStack" for pushes, "outStack" for pops/peeks.
 *           Transfer elements from inStack to outStack only when outStack
 *           is empty (amortized O(1)).
 * Time  : Amortized O(1) for push/pop/peek
 * Space : O(n)
 */

import java.util.Stack;

public class P232_ImplementQueueUsingStacks {

    static class MyQueue {
        private final Stack<Integer> inStack;
        private final Stack<Integer> outStack;

        public MyQueue() {
            inStack = new Stack<>();
            outStack = new Stack<>();
        }

        public void push(int x) {
            inStack.push(x);
        }

        public int pop() {
            transferIfNeeded();
            return outStack.pop();
        }

        public int peek() {
            transferIfNeeded();
            return outStack.peek();
        }

        public boolean empty() {
            return inStack.isEmpty() && outStack.isEmpty();
        }

        private void transferIfNeeded() {
            if (outStack.isEmpty()) {
                while (!inStack.isEmpty()) {
                    outStack.push(inStack.pop());
                }
            }
        }
    }

    // ---------- Simple manual test ----------
    public static void main(String[] args) {
        MyQueue queue = new MyQueue();
        queue.push(1);
        queue.push(2);
        System.out.println(queue.peek());  // 1
        System.out.println(queue.pop());   // 1
        System.out.println(queue.empty()); // false
    }
}

File: LeetCode/P232_ImplementQueueUsingStacks.cpp
#include <bits/stdc++.h>
using namespace std;

class MyQueue {
private:
    stack<int> inStack;
    stack<int> outStack;

    void transferIfNeeded() {
        if (outStack.empty()) {
            while (!inStack.empty()) {
                outStack.push(inStack.top());
                inStack.pop();
            }
        }
    }

public:
    MyQueue() {
    }

    void push(int x) {
        inStack.push(x);
    }

    int pop() {
        transferIfNeeded();
        int value = outStack.top();
        outStack.pop();
        return value;
    }

    int peek() {
        transferIfNeeded();
        return outStack.top();
    }

    bool empty() {
        return inStack.empty() && outStack.empty();
    }
};

int main() {
    MyQueue q;
    q.push(1);
    q.push(2);
    cout << q.peek() << "\n";
    cout << q.pop() << "\n";
    cout << boolalpha << q.empty() << "\n";
    return 0;
}
